snapshotter/demux/cache/cache.go

```diff
@@ -15,29 +15,173 @@ package cache
 
 import (
 	"context"
+	"fmt"
+	"net"
+	"sync"
+	"time"
 
 	"github.com/containerd/containerd/snapshots"
 	"github.com/firecracker-microvm/firecracker-containerd/snapshotter/demux/proxy"
+	"github.com/hashicorp/go-multierror"
 )
 
 // SnapshotterProvider defines a snapshotter fetch function.
 type SnapshotterProvider = func(context.Context, string) (*proxy.RemoteSnapshotter, error)
 
-// Cache defines the interface for a snapshotter caching mechanism.
-type Cache interface {
-	// Retrieves the snapshotter from the underlying cache using the provided
-	// fetch function if the snapshotter is not currently cached.
-	Get(ctx context.Context, key string, fetch SnapshotterProvider) (*proxy.RemoteSnapshotter, error)
+// RemoteSnapshotterCache implements a cache for remote snapshotters.
+type RemoteSnapshotterCache struct {
+	mutex        *sync.RWMutex
+	snapshotters map[string]*proxy.RemoteSnapshotter
 
-	// WalkAll applies the provided function across all cached snapshotters.
-	WalkAll(ctx context.Context, fn snapshots.WalkFunc, filters ...string) error
+	fetch SnapshotterProvider
 
-	// Closes the snapshotter and removes it from the cache.
-	Evict(key string) error
+	reaper *sync.Once
+	evict  chan string
+	lease  EvictionPolicy
 
-	// Releases the cache's internal resources and closes any cached snapshotters.
-	Close() error
+	stop chan struct{}
+}
+
+// SnapshotterCacheOption is a functional option that operates on a remote snapshotter cache.
+type SnapshotterCacheOption func(*RemoteSnapshotterCache)
+
+// NewRemoteSnapshotterCache creates a new instance with an empty cache and applies any provided caching options.
+func NewRemoteSnapshotterCache(fetch SnapshotterProvider, opts ...SnapshotterCacheOption) *RemoteSnapshotterCache {
+	c := &RemoteSnapshotterCache{
+		mutex:        &sync.RWMutex{},
+		snapshotters: make(map[string]*proxy.RemoteSnapshotter),
+		fetch:        fetch,
+		reaper:       &sync.Once{},
+		stop:         make(chan struct{}),
+	}
+
+	for _, opt := range opts {
+		opt(c)
+	}
+
+	return c
+}
+
+// EvictOnConnectionFailure is a caching option for evicting entries from the cache after a failed connection attempt.
+func EvictOnConnectionFailure(dial func(context.Context, string) (net.Conn, error), frequency time.Duration) func(*RemoteSnapshotterCache) {
+	return func(c *RemoteSnapshotterCache) {
+		c.evict = make(chan string)
+		c.lease = NewEvictOnConnectionFailurePolicy(c.evict, c.stop, dial, frequency)
+		c.startBackgroundReaper()
+	}
+}
+
+func (c *RemoteSnapshotterCache) startBackgroundReaper() {
+	reap := func() {
+		for {
+			s, ok := <-c.evict
+			if !ok {
+				break
+			}
+			c.Evict(s)
```

Comment:
Evict is returning an error. If that is not expected, we may need to either log the error or stop the reaper and tell the caller.

Reply:
Will be added in next revision.